### PR TITLE
Fix wrapping outputs. Replacement original forward for forward export.

### PR DESCRIPTION
**Pull request by asenina**
Fix 55333

### CODE REVIEW

**Comment by asenina** [comment]
please take a look @aleksei-kashapov

**Comment by kshpv** [comment]
After the discussion with @asenina we would like to have a discussion with someone from OTE side about this PR. 
The problem - NNCF initialization algorithm is blocked by a current realization of forward_export().

**Comment by druzhkov-paul** [comment]
@aleksei-kashapov Is one @LeonidBeynenson enough for you or do you need another one?

File: mmdet/models/detectors/base.py [hunk 1]
```diff
@@ -213,8 +213,10 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
         else:
             return self.forward_test(img, img_metas, **kwargs)
 
-    def forward_export(self, imgs):
+    def forward_export(self, imgs, img_metas=None):
         from torch.onnx.operators import shape_as_tensor
+        if img_metas is not None:
+                self.img_metas = img_metas
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py]:
```suggestion
    def forward_export(self, imgs, img_metas=None):
        if img_metas is not None and self.img_metas is not None:
            raise RuntimeError('forward_export is called with non-zero img_metas inside forward_export_context')
        from torch.onnx.operators import shape_as_tensor
        if img_metas is None:
                img_metas = self.img_metas
```

And, please, everywhere below replace `self.img_metas` with `img_metas`.
I think that this function should not set or replace the field `self.img_metas` -- IMHO it should be done in special places (I mean, such methods as `forward_export_context` and `forward_nncf_initialization_context`)

File: mmdet/models/detectors/base.py [hunk 2]
```diff
@@ -223,7 +225,10 @@ def forward_export(self, imgs):
         assert len(self.img_metas[0]) == imgs_per_gpu, f'self.img_metas={self.img_metas}'
         self.img_metas[0][0]['img_shape'] = img_shape[2:4]
 
-        return self.simple_test(imgs[0], self.img_metas[0], postprocess=False)
+        retval = self.simple_test(imgs[0], self.img_metas[0], postprocess=False)
+        if img_metas is not None:
+                self.img_metas = None
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py]: Please, remove this code here.
```suggestion
```

As I stated above, I think, the field `self.img_metas` should not be set or cleared in the function

File: mmdet/models/detectors/base.py [hunk 3]
```diff
@@ -223,7 +225,10 @@ def forward_export(self, imgs):
...
+        return retval
 
     @contextmanager
     def forward_export_context(self, img_metas):
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py:252]: I would recommend to use the following functions
```python
    @contextmanager
    def forward_nncf_initialization_context(self):
        prev_img_metas = self.img_metas
        self.img_metas = None
        forward_backup = self.forward
        self.forward = partial(self.forward, return_loss=False, forward_export=True)
        yield
        self.forward = forward_backup
        self.img_metas = prev_img_metas

    @contextmanager
    def forward_export_context(self, img_metas):
        prev_img_metas = self.img_metas
        self.img_metas = img_metas
        forward_backup = self.forward
        self.forward = partial(self.forward, return_loss=False, forward_export=True, img_metas=None)
        yield
        self.forward = forward_backup
        self.img_metas = prev_img_metas
```
Please, note that this implementations allow to use nested calls of the context managers `forward_export_context` and/or `forward_nncf_initialization` (one inside another should work)

File: mmdet/integration/nncf/compression.py [hunk 4]
```diff
@@ -200,7 +203,9 @@ def wrap_inputs(args, kwargs):
         kwargs['img'] = img
         return args, kwargs
 
+    original_forward = model.forward
     model.dummy_forward_fn = dummy_forward
+    model.forward = partial(model.forward, return_loss=False, forward_export=True)
```
Comment (LeonidBeynenson) [inline mmdet/integration/nncf/compression.py]: I would recommend you remove this version of changes.
Instead of that I would recommend to use the context received from the method `create_compressed_model` as follows:
```python
with model.forward_nncf_initialization_context():
    compression_ctrl, model = create_compressed_model(model,
                                                      nncf_config, ...)
```
See implementation of `forward_nncf_initialization_context` below

Please, note that using the context manager will allow you to avoid tricks with
storing and restoring `model.forward`. IMHO, it is more pretty.

**Review by LeonidBeynenson — CHANGES_REQUESTED** [review]
2 @asenina : Anastasia, I propose some changes. Please, use them -- I think they could improve the code a bit.

**Comment by kshpv** [comment]
> @aleksei-kashapov Is one @LeonidBeynenson enough for you or do you need another one?

We have discussed with @LeonidBeynenson, so it seems that a solution was found

File: mmdet/models/detectors/base.py (hunk 1 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: Done.

File: mmdet/models/detectors/base.py (hunk 2 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: Done.

File: mmdet/models/detectors/base.py (hunk 3 above)
Comment (asenina) [reply mmdet/models/detectors/base.py:252]: Done.

File: mmdet/integration/nncf/compression.py (hunk 4 above)
Comment (asenina) [reply mmdet/integration/nncf/compression.py]: Done.
@LeonidBeynenson  thanks for the help. So really prettier.

**Comment by asenina** [comment]
run tests

**Comment by asenina** [comment]
@LeonidBeynenson  please take another look and approve

File: mmdet/apis/train.py [hunk 5]
```diff
@@ -100,7 +100,17 @@ def train_detector(model,
...
+        data_loaders_for_init = [
+        build_dataloader(
+            ds,
+            1,
+            cfg.data.workers_per_gpu,
+            # cfg.gpus will be ignored if distributed
+            1,
+            dist=distributed,
+            seed=cfg.seed) for ds in dataset
+    ]
```
Comment (LeonidBeynenson) [inline mmdet/apis/train.py]: Please, fix indents in this code block

File: mmdet/models/detectors/base.py [hunk 6]
```diff
@@ -213,29 +213,31 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
         else:
             return self.forward_test(img, img_metas, **kwargs)
 
-    def forward_export(self, imgs):
+    def forward_export(self, imgs, img_metas=None):
+        if img_metas is not None and self.img_metas is not None:
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py:235]: Please, add below check that 
```
if img_metas is None and self.img_metas is None:
    raise RuntimeError(".....')
```

File: mmdet/models/detectors/base.py [hunk 7]
```diff
@@ -213,29 +213,31 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
         else:
             return self.forward_test(img, img_metas, **kwargs)
 
-    def forward_export(self, imgs):
+    def forward_export(self, imgs, img_metas=None):
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py:216]: Please, add pydoc comment -- something like
```
This function is called from the function `forward_test` in one of the cases: 
* During export of the model. In this case 
  * `forward_export_context` is used, 
  * `img_metas` is passed to the model using the field `self.img_metas` 
* During compression of the model by NNCF library, for building graph of 
  the model using   `dummy_forward` function. In this case 
  * `forward_export_context` is used too, 
  * also `img_metas` is passed to the model using the field `self.img_metas` 
* During compression of the model by NNCF library, for initialization of 
  the model. In this case
  * `forward_nncf_initialization_context` is used
  * `img_metas` is passed to the model through parameter of this function 

See the methods `forward_export_context`  and `forward_nncf_initialization_context`  below.
```

File: mmdet/models/detectors/base.py [hunk 8]
```diff
@@ -213,29 +213,31 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
-        self.img_metas[0][0]['img_shape'] = img_shape[2:4]
+        assert len(img_metas[0]) == imgs_per_gpu, f'img_metas={img_metas}'
+        img_metas[0][0]['img_shape'] = img_shape[2:4]
 
-        return self.simple_test(imgs[0], self.img_metas[0], postprocess=False)
+        retval = self.simple_test(imgs[0], img_metas[0], postprocess=False)
+        return retval
 
     @contextmanager
     def forward_export_context(self, img_metas):
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py:252]: Please, add to this function pydoc comment -- something like
```
This context is used for export of the model to ONNX and call the model
inside NNCF library to build the graph of the model. 
It does the following changes:
* it makes the method  `forward` of the model to call the method `forward_export` of the model
* it passes `img_metas` to the model through a field of the model to avoid 
  passing it through parameters of the function `forward`   (the latter is impossible in this case)
```

File: mmdet/models/detectors/base.py [hunk 9]
```diff
@@ -249,6 +251,16 @@ def forward_dummy_context(self, img_metas):
         self.forward_backup = None
         self.img_metas = None
 
+    @contextmanager
+    def forward_nncf_initialization_context(self):
```
Comment (LeonidBeynenson) [inline mmdet/models/detectors/base.py:283]: Please, add to this function pydoc comment -- something like
```
This context is used for calling the model inside NNCF library to initialize the NNCF inner variables using data from a dataset. 
It makes the method  `forward` of the model call the method 
`forward_export` of the model.
(It is required to make the model graph during NNCF initialization be the same
as the model graph during building of the graph.)
But (in contrast with `forward_export_context`) the `img_metas` is passed to the model through the parameter of the method `forward`.

**Review by LeonidBeynenson — CHANGES_REQUESTED** [review]

File: mmdet/apis/train.py (hunk 5 above)
Comment (LeonidBeynenson) [reply mmdet/apis/train.py]: And, please, do not put the `build_dataloader` into a list, -- as far as I can see, the only place when the list is used is the parameter `data_loaders_for_init[0]` , is not it?
so list is not required?

File: mmdet/models/detectors/base.py (hunk 6 above)
Comment (asenina) [reply mmdet/models/detectors/base.py:235]: Done.

File: mmdet/models/detectors/base.py (hunk 7 above)
Comment (asenina) [reply mmdet/models/detectors/base.py:216]: Done.

File: mmdet/models/detectors/base.py (hunk 8 above)
Comment (asenina) [reply mmdet/models/detectors/base.py:252]: Done.

File: mmdet/models/detectors/base.py (hunk 9 above)
Comment (asenina) [reply mmdet/models/detectors/base.py:283]: Done.

**Comment by asenina** [comment]
@LeonidBeynenson please approve

**Review by LeonidBeynenson — APPROVED** [review]

**Comment by LeonidBeynenson** [comment]
2 @druzhkov-paul : Pavel, could you review this PR?
It makes more complicated logic concerning the `forward_export` method, since now the method is  used for two different actions. See the comments that were added.
Is this a good decision in your opinion? Or another similar function be should be implemented for NNCF initialization only? 
If you find another option how to implement context for NNCF, it would be great.

File: mmdet/apis/train.py [hunk 10]
```diff
@@ -100,7 +100,17 @@ def train_detector(model,
     # nncf model wrapper
     nncf_enable_compression = bool(cfg.get('nncf_config'))
     if nncf_enable_compression:
-        compression_ctrl, model = wrap_nncf_model(model, cfg, data_loaders[0], get_fake_input)
+        data_loader_for_init = build_dataloader(
+            dataset[0],
+            1,
+            cfg.data.workers_per_gpu,
+            # cfg.gpus will be ignored if distributed
+            1,
```
Comment (druzhkov-paul) [inline mmdet/apis/train.py]: Will it be a problem to use `len(cfg.gpu_ids)` as for ordinary `data_loaders`?

File: mmdet/apis/train.py [hunk 11]
```diff
@@ -100,7 +100,17 @@ def train_detector(model,
...
+            dist=distributed,
+            seed=cfg.seed
+        )
+
+        compression_ctrl, model = wrap_nncf_model(model, cfg, data_loader_for_init, get_fake_input)
```
Comment (druzhkov-paul) [inline mmdet/apis/train.py:113]: Just out of curiosity, are both `data_loader_for_init` and `get_fake_input` required here? @LeonidBeynenson

File: mmdet/models/detectors/base.py [hunk 12]
```diff
@@ -213,29 +213,61 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
+        if img_metas is None and self.img_metas is None:
+            raise RuntimeError("forward_export is called with zero img_metas.")
+
+        if img_metas is not None and self.img_metas is not None:
+            raise RuntimeError('forward_export is called with non-zero img_metas inside forward_export_context.')
+
         from torch.onnx.operators import shape_as_tensor
-        assert self.img_metas, 'Error: forward_export should be called inside forward_export_context'
+        if img_metas is None:
+                img_metas = self.img_metas
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py]: Seems that indent is a bit too big.

File: mmdet/models/detectors/base.py [hunk 13]
```diff
@@ -213,29 +213,61 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
+    def forward_export(self, imgs, img_metas=None):
+        """
+        This function is called from the function `forward_test` in one of the cases:
+        * During export of the model. In this case
+          * `forward_export_context` is used,
+          * `img_metas` is passed to the model using the field `self.img_metas`
+        * During compression of the model by NNCF, for building graph of
+          the model using `dummy_forward` function. In this case
+          * `forward_export_context` is used too,
+          * also `img_metas` is passed to the model using the field `self.img_metas`
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py]: Two cases look identical. Can you please state explicitly that they are?

File: mmdet/models/detectors/base.py [hunk 14]
```diff
@@ -249,6 +281,27 @@ def forward_dummy_context(self, img_metas):
...
 
+    @contextmanager
+    def forward_nncf_initialization_context(self):
+        """
+        This context is used for calling the model inside NNCF to initialization compression algorithm
+        the NNCF using data from a dataset.
+        It makes the method  `forward` of the model call the method
+        `forward_export` of the model.
+        (It is required to make the model graph during NNCF initialization be the same
+        as the model graph during building of the graph.)
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py]: What?

File: mmdet/models/detectors/base.py [hunk 15]
```diff
@@ -213,29 +213,61 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
+          * `forward_export_context` is used,
+          * `img_metas` is passed to the model using the field `self.img_metas`
+        * During compression of the model by NNCF, for building graph of
+          the model using `dummy_forward` function. In this case
+          * `forward_export_context` is used too,
+          * also `img_metas` is passed to the model using the field `self.img_metas`
+        * During compression of the model by NNCF, for initialization of
+          the model. In this case
+          * `forward_nncf_initialization_context` is used
+          * `img_metas` is passed to the model through parameter of this function.
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py:227]: Why does the same flow as for export does not work for NNCF initialization?

Comment (LeonidBeynenson) [reply mmdet/models/detectors/base.py:227]: Pavel, the same flow cannot be used for NNCF initialization, since during initialization real elements from a dataloader are used.
Note that for export and for building NNCF graph (i.e. for dummy_forward function) some dummy random input is used -- the dummy input is generated once, its `img_meta` is also generated once, and the `img_meta` is stored in the model field.
For NNCF initialization real dataset elements are used with their `img_meta`-s. The `img_meta`-s may be different for different elements, so they cannot be stored as a field of the model.

**Comment by asenina** [comment]
run tests

File: mmdet/apis/train.py (hunk 10 above)
Comment (asenina) [reply mmdet/apis/train.py]: Done

File: mmdet/models/detectors/base.py (hunk 12 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: Fixed

File: mmdet/models/detectors/base.py (hunk 13 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: Done

File: mmdet/models/detectors/base.py (hunk 14 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: I wanted to say that the computational graph during tracing and building the model graph( in this case we use `dummy_forward`) and the computational graph during apply NNCF initialization algo (in this case we use the original forward `model(*args, **kwargs)` ) must be identical at the time of calling `create_compressed_model`.

Comment (druzhkov-paul) [reply mmdet/models/detectors/base.py]: Maybe I'm missing something, but phrases like "model graph during building of the graph" or "computational graph during building the model graph" make a little sense to me. What's the difference between model graph and computational graph?

**Comment by druzhkov-paul** [comment]
run tests

File: mmdet/apis/train.py (hunk 11 above)
Comment (LeonidBeynenson) [reply mmdet/apis/train.py:113]: Yes! They are used in different part of NNCF model wrapping process!

File: mmdet/models/detectors/base.py (hunk 14 above)
Comment (asenina) [reply mmdet/models/detectors/base.py]: This is the same.

Comment (druzhkov-paul) [reply mmdet/models/detectors/base.py]: So, please paraphrase the comment.

Comment (asenina) [reply mmdet/models/detectors/base.py]: `It is required during NNCF initialization call the same  forward function as during build the internal graph representation via tracing in NNCF.` is it good?

Comment (druzhkov-paul) [reply mmdet/models/detectors/base.py]: > It is required to call the same forward function during NNCF initialization and building of the internal NNCF graph representation via tracing.

I would go with this one, if it still keeps the semantics.

Comment (asenina) [reply mmdet/models/detectors/base.py]: Done.

**Review by druzhkov-paul — APPROVED** [review]

**Comment by druzhkov-paul** [comment]
As long as you (for some reason 🤷) do force-pushing, I have to run the tests one more time and wait until they successfully pass.

**Comment by druzhkov-paul** [comment]
run tests

**Comment by asenina** [comment]
@druzhkov-paul please merge
